BLE Indoor Positioning/src/test/java/com/nexenio/bleindoorpositioning/location/provider/LocationPredictorTest.java

```diff
@@ -0,0 +1,95 @@
+        Location predictedLocation = LocationPredictor.predict(subList, TimeUnit.SECONDS.toMillis(1));
+        double expectedDistance = lastKnownLocation.getDistanceTo(expectedLocation);
+        double actualDistance = lastKnownLocation.getDistanceTo(predictedLocation);
+        assertEquals(expectedDistance, actualDistance, 0.0001);
+    }
+
+    @Test
+    public void predictLocationFromLocations_locationsInCircle_accuratePrediction() throws Exception {
+        List<Location> walkInACircle = createWalkingCircle(GENDARMENMARKT);
+
+        for (int i = 0; i < walkInACircle.size() - 3; i++) {
+            List<Location> subList = walkInACircle.subList(i, i + 3);
+            Location lastKnownLocation = walkInACircle.get(i + 2);
+
+            Location expectedLocation = walkInACircle.get(i + 3);
+            Location predictedLocation = LocationPredictor.predict(subList, TimeUnit.SECONDS.toMillis(1));
+
+            // angle
+            double expectedAngle = lastKnownLocation.getAngleTo(expectedLocation);
+            double actualAngle = lastKnownLocation.getAngleTo(predictedLocation);
+            assertEquals(expectedAngle, actualAngle, 50);
+
+            // distance
+            double expectedDistance = lastKnownLocation.getDistanceTo(expectedLocation);
+            double actualDistance = lastKnownLocation.getDistanceTo(predictedLocation);
+            assertEquals(expectedDistance, actualDistance, 1);
+        }
+    }
+
+    @Test
+    public void calculateSpeed_locationsLine_correctSpeed() throws Exception {
+        List<Location> lineOnGendarmenmarkt = createWalkingLine(GENDARMENMARKT);
+        double speed = LocationPredictor.calculateSpeed(lineOnGendarmenmarkt);
+        assertEquals(speed, HUMAN_WALKING_SPEED, 0.1);
+    }
+
+    @Test
+    public void calculateSpeed_locationCircle_correctSpeed() throws Exception {
+        List<Location> walkInACircle = createWalkingCircle(GENDARMENMARKT);
+        for (int i = 0; i < walkInACircle.size() - 3; i++) {
+            List<Location> subList = walkInACircle.subList(i, i + 3);
+            double speed = LocationPredictor.calculateSpeed(subList);
+            assertEquals(speed, HUMAN_WALKING_SPEED, 0.1);
+        }
+    }
+
+    public List<Location> createWalkingLine(Location location) {
+        return createWalkingLocations(location, 0, 0, 10);
+    }
+
+    public List<Location> createWalkingCircle(Location location) {
+        return createWalkingLocations(location, 0, 360, 36);
+    }
+
+    public List<Location> createWalkingLocations(Location location, int startAngle, int stopAngle, int steps) {
+        long timestampDelta = 500;
+        double distanceToNextLocation = HUMAN_WALKING_SPEED * ((float) timestampDelta / TimeUnit.SECONDS.toMillis(1));
+        List<Location> walkInACircle = new ArrayList<>();
+        walkInACircle.add(location);
+        int angle = Math.abs(startAngle - stopAngle) / steps;
```

Comment:
This may result in rounding issues. This also seems to not work for e.g. `startAngle = 270` and `stopAngle = 90`